source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,320 @@
+    host = parent_.hosts_[rand_idx];
+  }
+
+  current_host_address_ = host->address()->asString();
+  RedisDiscoveryClientPtr& client = client_map_[current_host_address_];
+  if (!client) {
+    client = std::make_unique<RedisDiscoveryClient>(*this);
+    client->host_ = current_host_address_;
+    client->client_ = client_factory_.create(host, dispatcher_, *this);
+    client->client_->addConnectionCallbacks(*client);
+  }
+
+  current_request_ = client->client_->makeRequest(ClusterSlotsRequest::instance_, *this);
+}
+
+void RedisCluster::RedisDiscoverySession::onResponse(
+    NetworkFilters::Common::Redis::RespValuePtr&& value) {
+  current_request_ = nullptr;
+
+  // do nothing if the cluster is empty
+  if (value->type() != NetworkFilters::Common::Redis::RespType::Array || value->asArray().empty()) {
+    onUnexpectedResponse(value);
+    return;
+  }
+
+  std::vector<ClusterSlot> slots_;
+
+  // loop through the cluster slot response and error checks for each field
+  for (const NetworkFilters::Common::Redis::RespValue& part : value->asArray()) {
+    if (part.type() != NetworkFilters::Common::Redis::RespType::Array) {
+      onUnexpectedResponse(value);
+      return;
+    }
+    const std::vector<NetworkFilters::Common::Redis::RespValue>& slot_range = part.asArray();
+    if (slot_range.size() < 3 ||
+        slot_range[0].type() !=
+            NetworkFilters::Common::Redis::RespType::Integer || // start slot range as an integer
+        slot_range[1].type() !=
+            NetworkFilters::Common::Redis::RespType::Integer || // end slot range as an integer
+        slot_range[2].type() !=
+            NetworkFilters::Common::Redis::RespType::Array) { // master ip and port pair as an array
+      onUnexpectedResponse(value);
+      return;
+    }
+
+    // Field 2: Master address for slot range
+    // TODO(hyang): For now we're only adding the master node for each slot. When we're ready to
+    //  send requests to replica nodes, we need to add subsequent address in the response as
+    //  replica nodes.
+    auto master_address = ProcessCluster(slot_range[2]);
+    if (!master_address) {
+      onUnexpectedResponse(value);
+      return;
+    }
+    slots_.emplace_back(slot_range[0].asInteger(), slot_range[1].asInteger(), master_address);
+  }
+
+  parent_.onClusterSlotUpdate(slots_);
+  resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+}
+
+void RedisCluster::RedisDiscoverySession::onUnexpectedResponse(
+    const NetworkFilters::Common::Redis::RespValuePtr& value) {
+  ENVOY_LOG(warn, "Unexpected response to cluster slot command: {}", value->toString());
+  this->parent_.info_->stats().update_failure_.inc();
+  resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+}
+
+void RedisCluster::RedisDiscoverySession::onFailure() {
+  current_request_ = nullptr;
+  if (!current_host_address_.empty()) {
+    auto client_to_delete = client_map_.find(current_host_address_);
+    if (client_to_delete != client_map_.end()) {
```

Comment:
Can this ever not be true?

Reply:
It's possible if the remote connection was close and if the client's onEvent got invoked before the onFailure got called.  This is not possible with the current implementation, I put it in here just to be defensive.

Reply:
OK that's what I thought. Can you remove this error handling and anything similar that can't happen? Than you.

/wait